### PR TITLE
Uptycs integration update - add detections and queryJob api

### PR DESCRIPTION
**Pull request by cgadde-uptycs**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

1) Added new commands to the integration
2) One of the Uptycs api used in the integration was deprecated, all commands using that api were updated to use a different Uptycs api.
3) Testcases for new commands added and updated for modified commands.

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by cgadde-uptycs** [comment]
@YuvHayun Not sure what to do for the failing checks, made sure all commands are covered and tests are passing. Please review.

**Comment by YuvHayun** [comment]
Hey @cgadde-uptycs, I've sent you an email to set a meeting to discuss the pr.
If you can't find the mail please reach me on the Slack's DFIR channel (Yuval Hayun) or on email yhayun@paloaltonetworks.com

**Comment by cgadde-uptycs** [comment]
@YuvHayun Saw your mail, we can get into a call to discuss. Also let me update the PR description with details.

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.py [hunk 1]
```diff
@@ -367,12 +451,166 @@ def uptycs_get_assets_command():
...
+    severity = demisto.args().get('severity')
+    if severity != -1 and severity is not None:
+        filters.append('"severity":{"in":["%s"]}' % severity)
+
+    status = demisto.args().get('status')
+    if status != -1 and status is not None:
+        filters.append('"status":{"in":["%s"]}' % status)
+
+    asset_type = demisto.args().get('asset_type')
+    if asset_type != -1 and asset_type is not None:
```
Comment (YuvHayun) [inline Packs/Uptycs/Integrations/Uptycs/Uptycs.py:480]: Why would we expect to get -1 here?

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.py [hunk 2]
```diff
@@ -367,12 +451,166 @@ def uptycs_get_assets_command():
...
+def uptycs_get_detections():
+    """
+    return list of detections
+    """
+    http_method = 'get'
+    api_call = "/detections"
+
+    query_parameters = []
+    limit = demisto.args().get('limit')
+    if limit != -1 and limit is not None:
```
Comment (YuvHayun) [inline Packs/Uptycs/Integrations/Uptycs/Uptycs.py:463]:
```suggestion

    args = demisto.args()
    query_parameters = []
    limit = args.get('limit')
    if limit != -1 and limit is not None:
```

Let's switch this for all other places in the function that call demisto.args()

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.py [hunk 3]
```diff
@@ -367,12 +451,166 @@ def uptycs_get_assets_command():
...
+    query_results = uptycs_get_detection_details()
+    detection = query_results.get('detection')
+    final_alerts = []
+    detection_url = 'https://' + DOMAIN + '/ui/detections/' + demisto.args().get('detection_id')
+    display_name = get_href_link(detection["displayName"], detection_url)
+
+    for alert in query_results.get('alerts').get('items'):
+        alert_text = alert['alertTime'] + '<br>Alert - ' + alert['alertRuleName']
+        alert_text += '<br>\tscore - ' + alert['score'] + '<br>\t'
+        alert_text += alert['key'] + ' : ' + alert['value'] + '<br>\t'
```
Comment (YuvHayun) [inline Packs/Uptycs/Integrations/Uptycs/Uptycs.py:554]:
```suggestion
        alert_text += alert.get('key', "") + ' : ' + alert.get('value', "") + '<br>\t'
```
Let's refrain from direct access to dict elements (for all other locations in the code).

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.py [hunk 4]
```diff
@@ -367,12 +451,166 @@ def uptycs_get_assets_command():
...
+def uptycs_get_detection_details_command():
+    query_results = uptycs_get_detection_details()
+    detection = query_results.get('detection')
+    final_alerts = []
+    detection_url = 'https://' + DOMAIN + '/ui/detections/' + demisto.args().get('detection_id')
+    display_name = get_href_link(detection["displayName"], detection_url)
+
+    for alert in query_results.get('alerts').get('items'):
+        alert_text = alert['alertTime'] + '<br>Alert - ' + alert['alertRuleName']
+        alert_text += '<br>\tscore - ' + alert['score'] + '<br>\t'
```
Comment (YuvHayun) [inline Packs/Uptycs/Integrations/Uptycs/Uptycs.py:553]: why using <br>? you can use \n.

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.py [hunk 5]
```diff
@@ -367,12 +451,166 @@ def uptycs_get_assets_command():
...
+    context = {'id': detection['id'],
+               'displayName': display_name,
+               'score': detection['score'],
+               'severity': detection['severity'],
+               'createdAt': detection['createdAt'],
+               'signals': detection['signals'],
+               'tacticCount': detection['tacticCount'],
+               'tactics': str(detection['tactics']),
+               'assetHostName': detection['assetHostName'],
+               'Alerts and Events': '<br>'.join(sorted(final_alerts)),
```
Comment (YuvHayun) [inline Packs/Uptycs/Integrations/Uptycs/Uptycs.py:581]: why do we want to retrieve all the text in the context?
context is better for objects where we can retrieve the information from and move with it to the next task.
I think it's better to create a list of dicts with the info and use the TableToMarkdown.

**Review by YuvHayun — CHANGES_REQUESTED** [review]
Hey @ cgadde-uptycs, thanks a lot for your contribution!
I've left some comments that needs to be fixed.
Let me know when you're ready to re-review.
Feel free to reach me if any question comes up.

**Comment by YuvHayun** [comment]
Hey @cgadde-uptycs how's it going?
Let me know if you need any help with the pr.

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.yml [hunk 6]
```diff
@@ -1518,7 +1518,170 @@ script:
     - name: filename
       required: true
       description: The name of the file being uploaded. This file should be uploaded to Cortex XSOAR in the Playground War Room using the paperclip icon next to the CLI.
-  dockerimage: demisto/auth-utils:1.0.0.79567
+  - name: uptycs-get-detections
+    description: Get list of detections from Uptycs,
+    arguments:
+    - name: limit
+      description: Limit the number of entries returned.  Use -1 to return all entries (may run slow or cause a time out).
+      defaultValue: "10"
```
[the remark this answers is not in the document]
Comment (cgadde-uptycs) [reply Packs/Uptycs/Integrations/Uptycs/Uptycs.yml]: updated as suggested

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.yml [hunk 7]
```diff
@@ -1518,7 +1518,170 @@ script:
...
+    - name: asset_type
+      auto: PREDEFINED
+      predefined:
+      - asset
+      - cloud
+      - kubernetes
+      description: Asset Type of the detection.
+      type: textArea
+    - name: time_ago
+      description: 'Specifies how far back you want to look.  Format examples: 2 hours, 4 minutes, 6 month, 1 day, etc.'
```
[the remark this answers is not in the document]
Comment (cgadde-uptycs) [reply Packs/Uptycs/Integrations/Uptycs/Uptycs.yml]: updated as suggested

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.py [hunk 8]
```diff
@@ -221,27 +248,79 @@ def uptycs_parse_date_range(timeago, start_time, end_time):
...
 
+def uptycs_get_upt_day(start_time):
+    original_datetime = datetime.strptime(start_time, "%Y-%m-%d %H:%M:%S.%f")
+    return original_datetime.strftime("%Y%m%d")
+
+
 """COMMAND FUNCTIONS"""
 
 
+def uptycs_poll_queryjob_status(job_id, query):
```
[the remark this answers is not in the document]
Comment (cgadde-uptycs) [reply Packs/Uptycs/Integrations/Uptycs/Uptycs.py]: @YuvHayun I tried using the polling command function, but it was not waiting till the query completion, can you please check what I am missing ?

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.py [hunk 9]
```diff
@@ -221,27 +248,79 @@ def uptycs_parse_date_range(timeago, start_time, end_time):
...
+        offset += limit
+        if output.get('items'):
+            rows = output.get('items')
+            final_output.extend(remove_context_entries(rows, ['rowData']))
+            if len(rows) < limit:
+                return [entry['rowData'] for entry in final_output]
+        else:
+            # unknown or empty output
+            return_error("Invalid/Empty query job output")
+    return
```
[the remark this answers is not in the document]
Comment (cgadde-uptycs) [reply Packs/Uptycs/Integrations/Uptycs/Uptycs.py:309]: removed unnecessary statement.

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.py (hunk 1 above)
Comment (cgadde-uptycs) [reply Packs/Uptycs/Integrations/Uptycs/Uptycs.py]: removed unnecessary conditions.

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.py (hunk 2 above)
Comment (cgadde-uptycs) [reply Packs/Uptycs/Integrations/Uptycs/Uptycs.py]: updated as suggested

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.py (hunk 4 above)
Comment (cgadde-uptycs) [reply Packs/Uptycs/Integrations/Uptycs/Uptycs.py]: "<br>" is needed as I want to display the human readable output in tabular format.

File: Packs/Uptycs/Integrations/Uptycs/Uptycs.py (hunk 5 above)
Comment (cgadde-uptycs) [reply Packs/Uptycs/Integrations/Uptycs/Uptycs.py]: removed redundant statements and updated code to use context properly.

File: Packs/Uptycs/ReleaseNotes/1_0_12.md [hunk 10]
```diff
@@ -0,0 +1,22 @@
+
+#### Integrations
+##### Uptycs
+- Updated uptycs-run-query command code.
```
[the remark this answers is not in the document]
Comment (cgadde-uptycs) [reply Packs/Uptycs/ReleaseNotes/1_0_12.md]: updated as suggested

**Comment by cgadde-uptycs** [comment]
@YuvHayun Sorry for the delay. Addressed your comments. Can you please check the polling function ?

**Comment by YuvHayun** [comment]
Hey @cgadde-uptycs, how's it going? feel free to reach me (you already have my email) if you need any help.

**Comment by cgadde-uptycs** [comment]
@YuvHayun As discussed last time, since polling function is meant to be used for longer waiting periods, I reverted it back to using sleep, Maximum allowed wait time is only 120 secs. All other comments are addressed. Please check.

**Comment by YuvHayun** [comment]
Hey @cgadde-uptycs, as I explained during the call.. sleep is not allowed in the code.
I've sent you an email a week ago with an example of how to preform the poll command correctly.
Let me know if you can't find it/you need help to implement it.

**Comment by YuvHayun** [comment]
Hey @cgadde-uptycs, still waiting for an update on the poll commands.. let me know how's it going.

**Comment by YuvHayun** [comment]
Closing the pr for now.
Feel free to re-open when you have time to advance with it.

**Pull request closed without merging**